Try all galaxy servers when locating available versions for a collection

s-hertel opened the pull request:
##### SUMMARY
We handle GalaxyError in `get_collection_version_metadata` properly, but not in `get_collection_versions` which is called before getting a specific version's metadata for install/download.

I also made non-HTTPError exceptions a warning instead of fatal for collection install/download/verify (which probably needs discussion).  This would fix #75443.

##### ISSUE TYPE
- Bugfix Pull Request

s-hertel committed:
If a GalaxyError exception occurs when getting a collection's metadata, continue to the next in the server list.

s-hertel committed:
Warn for unknown exceptions when finding versions of a collection

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -52,13 +57,31 @@ def get_collection_versions(self, requirement):
+                display.warning(
+                    "Skipping Galaxy server %s. "
+                    "Got an unexpected error when getting "
+                    "available verisons of collection %s.%s: %s"
```
```suggestion
                    "available versions of collection %s.%s: %s"
```

s-hertel committed:
Update lib/ansible/galaxy/collection/galaxy_api_proxy.py

Co-authored-by: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -52,13 +57,31 @@ def get_collection_versions(self, requirement):
+                    "available verisons of collection %s.%s: %s"
+                    % (
+                        api.api_server,
+                        requirement.namespace,
```
Just use `requirement.fqcn` instead of concatenating ns+name manually.

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -52,13 +57,31 @@ def get_collection_versions(self, requirement):
+                display.warning(
+                    "Skipping Galaxy server %s. "
+                    "Got an unexpected error when getting "
+                    "available versions of collection %s.%s: %s"
```
Nit: it is preferable to use named placeholders when interpolating strings to improve maintainability and reduce mental overhead of having to enumerate the placeholders and map them to a separate list of values.

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -52,13 +57,31 @@ def get_collection_versions(self, requirement):
+        last_err = None
+        for api in api_lookup_order:
+            try:
+                for version in api.get_collection_versions(requirement.namespace, requirement.name):
```
I feel like is becoming complicated. I suggest you explore refactoring a part of this into a helper generator only wrapping with try/except a specific invention as opposed to the block with a few levels of nesting. That is if you feel like exploring this, it's not a blocker.

s-hertel replied:
I think it's better now with f4df143. The cyclomatic complexity was 8, now it's 5.

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -52,13 +57,31 @@ def get_collection_versions(self, requirement):
+                        requirement.name,
+                        to_text(err),
+                    )
+                )
```
Have you considered tracking these exceptions too?
```suggestion
                )
                last_err = err
```

s-hertel replied:
I did, but wasn't sure. I thought maybe we should show the last legitimate-looking error...

webknjaz replied:
Maybe. I don't have a strong opinion here. Maybe to use this exception too we need to count that there were no successful loop iterations (legitimately returning empty results).

s-hertel committed:
Fix specifying the server for verify tests to compare checksums against the same server used by the install task

s-hertel committed:
Prepend an inaccessible server to the ansible-galaxy-collection servers list

Installing a collection isn't fatal as long as it exists on a subsequent server

s-hertel committed:
Use requirement.fqcn

Use named placeholders in warning message string

Track Exception in addition to GalaxyError exceptions and only reraise the final error if no servers were successful

s-hertel committed:
Simplify by breaking logic into two helper methods

s-hertel committed:
minimize diff

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -35,6 +40,46 @@ def __init__(self, apis, concrete_artifacts_manager):
 
+    def _get_collection_versions_with_api_order(self, requirement, api_lookup_order):
+        # type: (Candidate, Iterable[GalaxyAPI]) -> Iterable[Tuple[GalaxyAPI, str]]
+        """Helper for get_collection_versions. Yield api, version pairs for all APIs and reraise the last error if no valid API was found."""
```
```suggestion
        """Helper for get_collection_versions.

        Yield (api, version) pairs for all APIs,
        and reraise the last error if no valid API was found.
        """
```

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -35,6 +40,46 @@ def __init__(self, apis, concrete_artifacts_manager):
         self._concrete_art_mgr = concrete_artifacts_manager
 
+    def _get_collection_versions_with_api_order(self, requirement, api_lookup_order):
+        # type: (Candidate, Iterable[GalaxyAPI]) -> Iterable[Tuple[GalaxyAPI, str]]
```
```suggestion
        # type: (Requirement, Iterable[GalaxyAPI]) -> Iterator[Tuple[GalaxyAPI, str]]
```

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -35,6 +40,46 @@ def __init__(self, apis, concrete_artifacts_manager):
+                )
+            )
+            error = unknown_err
+        return result, error
```
I don't like an API where there's a need for the caller [not to forget to check for a return code](https://mobile.twitter.com/jxxf/status/959490034163269632). It's something that is present in older C-like languages because of the limitations they have. But in Python it should follow EAPF.

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -35,6 +40,46 @@ def __init__(self, apis, concrete_artifacts_manager):
+        for api in api_lookup_order:
+            versions, error = self._get_collection_versions_with_api(requirement, api)
+
+            if error is None:
```
I think this should be done via try/except rather than "don't-forget-about-the-return-code" driven style...

s-hertel replied:
There's just one helper method now, which uses the original try/except flow rather than "don't-forget-about-the-return-code".

webknjaz replied:
Perfect!

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -35,6 +40,46 @@ def __init__(self, apis, concrete_artifacts_manager):
+        """Helper for get_collection_versions. Yield api, version pairs for all APIs and reraise the last error if no valid API was found."""
+        found_api = last_error = None
+
+        for api in api_lookup_order:
```
Looking closer, I wonder why this doesn't use `self._apis`. If it's different, then this method definitely doesn't belong in this class.

s-hertel replied:
Fixed.

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -35,6 +40,46 @@ def __init__(self, apis, concrete_artifacts_manager):
+            raise last_error
+
+    def _get_collection_versions_with_api(self, requirement, api):
+        # type: (Candidate, GalaxyAPI) -> Tuple[Optional[list], Optional[Exception]]
```
It's not a candidate according to how it's being called. Also, the lowercase `list` is only properly available in typing since Python 3.9 (or 3.10) + it should have some information on what it contains.
```suggestion
        # type: (Requirement, GalaxyAPI) -> Tuple[Optional[List[Any]], Optional[Exception]]
```

s-hertel committed:
Add tests for verify and make that code path mirror install/download behavior

Make requested changes

s-hertel committed:
Fix warning for verify subcommand

s-hertel committed:
changelog

s-hertel committed:
edit changelog

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -35,6 +40,47 @@ def __init__(self, apis, concrete_artifacts_manager):
         self._concrete_art_mgr = concrete_artifacts_manager
 
+    def _get_collection_versions(self, requirement):
+        # type: (Requirement, Iterator[GalaxyAPI]) -> Iterator[Tuple[GalaxyAPI, str]]
```
It seems like this change (and maybe some of the previous ones) break the typing check (which is still only present in an unmerged PR atm). I'll try to look into this more closely.

s-hertel replied:
:eyes:

webknjaz replied:
Oh, I see. It describes two args while accepting only one. It doesn't accept an iterator returning galaxy apis.

s-hertel replied:
Opened a PR if you want to review #76296

webknjaz replied:
Oh, I was working on my own patch today.

s-hertel replied:
Ah, okay. I took the comment as a cue that you wanted me to fix it.
